### ISSUE TITLE
desktop-windows context missing when using custom config file

### ISSUE THREAD

**Issue by karolz-ms** [docker/for-win#11530]
## Environment
- Windows 10 Enterprise Version 21H1 build 19043.1052
- Docker Desktop 3.5 build 66024 running in Windows container mode

## Repro steps
1. Create a `config.json` file with the following content

```json
{
  "auths": {
    "https://index.docker.io/v1/": {}
  },
  "credsStore": "desktop",
  "currentContext": "desktop-windows",
  "HttpHeaders": {
    "X-Meta-Source-Client": "vs/web/publish"
  }
}
```

2. Do `docker context ls`. Note that `default` and `desktop-windows` contexts are present and the `desktop-windows` context is the current context
3. Do `docker --config config.json context ls`

**Expected**
The output should be the same as the result of the previous command

**Actual**
Only the `default` context is listed (as current)

## More info/context
This might seem like a minor issue, but we (Microsoft) are facing a problem in Microsoft Visual Studio where the "missing" `desktop-windows` blocks publishing containerized .NET applications to Azure Container Registry or Azure App service. Unfortunately we could not reproduce this issue outside of Visual Studio so far. Repro steps below (using latest Visual Studio 2019)

1. Create an ASP.NET Core web project with docker support enabled ("choose Windows containers").
2. Right click on the project -> "publish", select "Azure" -> select "Azure Container Registry", click "Next" button, select an Container Registry then click "Finish" button.
3. After the application builds successfully, click "publish" button on the page that shows up.

** Expected **
The application should publish to Azure Container Registry.

** Actual **
The image tagging step fails with the following error

`Current context "desktop-windows" is not found on the file system, please check your config file at c:\temp\bea1054e3fc64c2d98fb4926b2b7e5a7\config.json`

The config file mentioned is generated by Visual Studio and looks exactly like the one above. We are pretty sure that "context not found on the file system" error message is coming from Docker.

**Comment by karolz-ms** [comment]
Update: we are facing the same issue with Docker running in Linux containers mode as well

**Comment by StefanScherer** [comment]
Thanks @karolz-ms for the report and the repro steps.
We're looking into it now.

**Comment by StefanScherer** [comment]
Please be aware that the option `--config` does not take a file, but only a folder where the client config file (and contexts, ...) are located.

```
Options:
      --config string      Location of client config files (default
                           "C:\\Users\\Stefan Scherer\\.docker")
```

Other contexts that are created in my default config location also won't show up when I use the `--option` flag. This is because it tries to find a folder with the name `config.json` and silently ignores that there is nothing.

But I agree, the `desktop-windows` and `desktop-linux` contexts are like the "default" context and have no file and therefore should probably show up here. I'll discuss with the team.

**Comment by karolz-ms** [comment]
@StefanScherer thanks I did not realize that about the `--config` option. How about `DOCKER_CONFIG` environment variable? Does it work the same way (i.e. is supposed to point to a folder, and not a single file)?

**Comment by karolz-ms** [comment]
When I set `DOCKER_CONFIG` to a folder with the config file described in the repro steps, and issue a Docker command like `docker tag` I am getting the error that we see in Visual Studio:

```shell
C:\code\scratch>type config.json
{
  "auths": {
    "https://index.docker.io/v1/": {}
  },
  "credsStore": "desktop",
  "currentContext": "desktop-windows",
  "HttpHeaders": {
    "X-Meta-Source-Client": "vs/web/publish"
  }
}

C:\code\scratch>docker context ls
Current context "desktop-windows" is not found on the file system, please check your config file at C:\code\scratch\config.json

C:\code\scratch>set DOCKER_CONFIG=

C:\code\scratch>docker context ls
NAME                TYPE                DESCRIPTION                               DOCKER ENDPOINT
    KUBERNETES ENDPOINT                                                                         ORCHESTRATOR
acictx              aci                 rg-aci2@westus

default             moby                Current DOCKER_HOST based configuration   npipe:////./pipe/docker_engine
       swarm
desktop-windows *   moby                                                          npipe:////./pipe/dockerDesktopWindowsEngine

C:\code\scratch>type c:\Users\karolz\.docker\config.json
{"auths":{"https://index.docker.io/v1/":{}},"credsStore":"desktop","currentContext":"desktop-windows"}
```

I assume this represents a good repro of the problem?

**Comment by Backrightup2113** [comment]
DevOps is a new trend in software development over the past few years. Although the term is relatively new, it is actually a combination of many practices that have been in use for years. DevOps allows software to be released quickly and efficiently, while still maintaining high levels of security. for more visit [Azure devops backup](https://backrightup.com/)

**Comment by StefanScherer** [comment]
@karolz-ms I think the workaround for now is to use the `default` context in custom config files for now. The default context is still valid in parallel to the newly introduced ones.
Could you explain why you need a custom location in Visual Studio?
We're currently looking into it to come up with a fix, but we want to understand how the configs are used by other products.

**Comment by djs55** [comment]
@karolz-ms I've got a developer build of Docker Desktop which still has `desktop-*` contexts, but switches users back to the `default` context on upgrade. Hopefully this fixes the issue. If you have a chance to try it, please let me know how it goes! https://desktop-stage.docker.com/win/stable/amd64/66318/Docker%20Desktop%20Installer.exe

**Comment by karolz-ms** [comment]
@StefanScherer the config is used solely to add a custom header to registry requests, so we can track the usage of Visual Studio "publish to Azure AppService" feature (refer to `X-Meta-Source-Client` above). 

The code in VS copies the Docker configuration file that represents the current context, then adds the header information. That is why it started leveraging the new contexts immediately after they were introduced. It is designed to use the current ambient Docker context transparently.

@djs55 thank you 🤗  -- will try the new build shortly and report back ASAP

**Comment by karolz-ms** [comment]
@djs55 this build does solve the main issue of publishing to Azure Container Registry from Visual Studio. Ship it! 😄 

I noticed though that the original problem still persists. In the new build the default `config.json` has only setting for `credStore`, it does not refer to context name at all. That makes most of the Docker commands work with custom config, including the ones that Visual Studio uses. However, contexts other than default one "disappear". So, for example, `docker context inspect somenondefaultctx` will fail.

**Comment by djs55** [comment]
@karolz-ms the patch with the fix is in our merge queue, so hopefully it'll be shipped soon!

Looking at the original issue description, I think if Visual Studio is making a temporary `config.json` with the configuration it needs (which is fine in principle), it might be missing the implementation details of the contexts (URLs etc) because they're stored in subdirectories, named by the hash of the context name, e.g.

```
% echo -n desktop-linux | shasum5.30 -a 256
fe9c6bd7a66301f49ca9b6a70b217107cd1284598bfc254700c989b916da791e  -

% ls ~/.docker/contexts 
meta

% ls ~/.docker/contexts/meta 
1bf417bd87162d5d77a5ce3259e81bdee9f8a0ef56bc0f12eadcd5805a94b708	7347a7c3638e040491bad4e62f13fc93342b86934585d99ab49a532c52551863
593787d465b2d4cb19caeb74abbad57ec712f5605620bb6c35f4a162ef06a2d4	b1c648f3ccdc2321795e00df9b5cefdd4f80dcddce3003a260bea734ba522bc5
599c6b0587cb351600af23f399b2162a7f91917b994f92edb7a7580db7e50f92	dbc375ec425d5abbe4b2fb49ebde7674a603eb908349264442db451e4133d47b
68693d02ab4fbb2331b8cc39915322e48e61f06d4d1b31e7d19913202857bc8a	fe9c6bd7a66301f49ca9b6a70b217107cd1284598bfc254700c989b916da791e

% ls ~/.docker/contexts/meta/fe9c6bd7a66301f49ca9b6a70b217107cd1284598bfc254700c989b916da791e
meta.json
```

I wonder if we need a command like `docker context clone <dir>`, similar to `git clone`, to make a snapshot of the current CLI config? (In my ideal world I would store the CLI config in `git`, but that's probably not something we can impose on everyone unfortunately!)

**Comment by karolz-ms** [comment]
@djs55 yes, I think having something like `docker context clone` would make sense.

I am re-reading the [configuration file documentation](https://docs.docker.com/engine/reference/commandline/cli/#configuration-files) and from that reading is not super clear to me that the whole context configuration folder needs to be cloned for the custom `config.json` file to work properly (or what exactly needs to be cloned, for that matter, since what is stored in the `.docker` folder is assumed to be implementation detail). So it seems like currently we are not quite setting up users for success with regards to custom configuration files.

**Comment by lorenrh** [comment]
We have just released Docker Desktop 3.5.2, containing a fix for this issue.
